### ISSUE TITLE
Representation of `Process` arguments

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#14773]
`Process.new`  (and `.run`) has two parameters, `command` for the command to be executed, and `args` for the arguments.

I'm not quite sure what's the reasoning behind this separation. It's been established since the early days of Crystal (https://github.com/crystal-lang/crystal/commit/0b35fa5d688110d898275c9ab145cce93a4faffc) and is probably inherited from Ruby's [`Kernel#exec`](https://ruby-doc.org/3.2.0/Kernel.html#method-i-exec).

In many (most?) other APIs, the equivalent method receives a single list of arguments, where the first one represents the command to execute. This is similar to the way process execution is modeled in Unix.

`Process` even uses this representation internally: The `command` and `args` parameters are merged into a single array in `Process.prepare_args`. Even on Windows.
This internal representation as `Array(String)` is additional allocation which is actually just an intermediary: On Unix, the array is mapped to an array of pointers to those strings. On Windows it's mapped to an array of wide strings (`Array(Slice(UInt16))`).
This intermediate representation should be avoidable.

However, we might also want to consider whether the public API could use the single array model where the command is just the first element. I'm not sure if this has any clear benefit expect being more similar to other APIs.

This was breviously mentioned in #9030 before. And I suppose this might have some relevancy the `shell: true` conundrum.

**Comment by ysbaddaden** [comment]
Well, on UNIX `exec(2)` distinguishes the executable path from the arg list.

Weirdly, `CreateProcessW` on Windows does the same, but in order to search path it must be NULL, and the command be merged with the rest of the args :confused:

**Comment by straight-shoota** [comment]
Receiving an array including the command would make it very convenient to use percent-like array literals for the entire command array. It reads just like a shell command, but it's actually an array:

```cr
# single array:
Process.run(%w[crystal tool format])
# equivalent to
Process.run(["crystal", "tool", "format"])

# reads nicer than
Process.run("crystal", ["tool", "format"])
Process.run("crystal", %w[tool format])
```

Percent-like array literals however do _not_ support interpolation. Thus, it's unfortunately only that convenient if the command is entirely static. We cannot inject runtime values into the array.

Since many command invocations consist of a static portion followed by some dynamic arguments we can use a workaround and append to the array:

```cr
# One argument
Process.run(%w[crystal tool format] << path)

# Multiple arguments
Process.run(%w[crystal tool format].concat(paths))

# Without percent-like literal
Process.run(["crystal", "tool", "format", path])
Process.run(["crystal", "tool", "format", *paths])
```

This works well enough for simple use cases. But it can become clunky when there are more arguments.
I'd reckon it still reads much better than a shell command with lots of `Process.quote` injections.

```cr
# shell command
Process.run("git ls-tree -r --full-tree --name-only #{Process.quote(ref.to_git_ref)} -- #{Process.quote(path)}", shell: true)

# command array
Process.run(%w[git ls-tree -r --full-tree --name-only] << ref.to_git_ref << "--" << path)
```

**Comment by straight-shoota** [comment]
For improved convenience maybe we could introduce a `%W` literal with support for interpolation?
```cr
Process.run(%W[git ls-tree -r --full-tree --name-only #{ref.to_git_ref} -- #{path}])
```
Ruby has `%W` literals: https://docs.ruby-lang.org/en/4.0/syntax/literals_rdoc.html#label-25w+and+-25W-3A+String-Array+Literals
Not sure why Crystal never adopted them...

Perhaps we could also consider splatting? (which Ruby apparently does not)
```cr
Process.run(%W[crystal tool format #{*paths}])
```

**Comment by straight-shoota** [comment]
I created an RFC for `%W` syntax: https://github.com/crystal-lang/rfcs/pull/21

**Comment by straight-shoota** [comment]
A limitation of using percent string literals for command line arguments is that escaping is different from what you would typically expect in a shell. Instead of wrapping a couple of words inside quotes, you would need to escape every single whitespace.

```cr
# shell:
`foo "bar baz"`
# string array:
%w[foo bar\ baz]
```

A perhaps more convenient alternative, especially for longer strings is interpolation. So again, interpolations are an improvement:
```cr
# string array with interpolation:
%W[foo #{"bar baz"}]
```

**Comment by ysbaddaden** [comment]
I'll say it again: there really is nothing wrong with an explicit `["foo", "bar baz"]` array literal, quite the opposite: explicit behavior, safe argument injection, and avoids stupid errors (oops):

```crystal
%w[foo 'bar baz'] # => ["foo", "'bar", "baz'"]
%w[foo "bar baz"] # => ["foo", "\"bar", "baz\""]
```

Interpolation in string array literals will be nice to have, but string array literals are still tricky in practice.

More examples:

```crystal
%W[foo bar#{"baz"}] # => ["foo", "barbaz"]
#         ^-- missing space (might be intended)

["foo", "bar" "#{"baz"}"] # => parse error!
#            ^-- missing comma (not intended)
```

**Comment by oprypin** [comment]
Good point actually. Let me add another point. There already exists an alternative to `%W` literals - just write `split`.

Both of these are the same mistake but only one of them stands out as such: 

```crystal
%W[foo 'bar baz']
"foo 'bar baz'".split
```

If `%W` has the ability to *seem* totally safe then it should actually *be* totally safe for this usecase, but it could never be.

**Comment by straight-shoota** [comment]
Of course, it's possible to shoot yourself in the foot. And I agree that the similar notation might encourage such mistakes.
I don't think it's a critical concern, though.

It only affects static parts of the literal. Interpolated parts in `%W` can't be split.
This should be easy to spot in a linter. And errors should show up relatively quickly.

That would be different with `"foo '#{bar}'".split` which works entirely fine as long as `bar` contains no whitespace.

**Comment by ysbaddaden** [comment]
`%W` shouldn't be meant for shell-like commands, it's an improvement over `%w`.

Both happen to improve readability as spawn args but also for other cases, but warning: it's tricky! Yet, not unsafe as in a malicious user can do harm: the dangerous part is the handling of user data, and that will be correctly quoted. _(edit: maybe merging with the previous or next arg could be exploited 🤔)_
